[HUDI-6501] HoodieHeartbeatClient should stop all heartbeats and not delete heartbeat files for close

SteNicholas opened the pull request:
### Change Logs

`HoodieHeartbeatClient` should close without stopping all heartbeats started via this instance of the client. Because the start of the heartbeat is executed in `startCommit`, therefore the stop of the heartbeat should be executed in `postCommit`. It doesn't need to stop all heartbeats started via this instance of the client when closing the `HoodieHeartbeatClient`.

For example, there is a timeline generated by Flink job: deltacommit1.inflight(Should be recommitted), deltacommit2.inflight. Meanwhile, there is a clustering job with clean using lazy failed writes clean policy. The Flink job failed with stopping heartbeat client, then the heartbeat files are deleted. At this time, the clustering job could rollback the deltacommit1, deltacommit2. After the Flink job restarts, the deltacommit1 is rollback and could not recommit.

In summary, the heartbeat file is created in `startCommit` and should be deleted in `postCommit` or `rollbackFailedWrites`, not in the close of the heartbeat client.

### Impact

HoodieHeartbeatClient closes the client without stopping all heartbeats started via this instance of the client.

### Risk level (write none, low medium or high below)

none.

### Documentation Update

none.

### Contributor's checklist

- [x] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [x] Change Logs and Impact were stated clearly
- [x] Adequate tests were added if applicable
- [x] CI passed

stream2000 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -201,14 +196,6 @@ public void stop(String instantTime) throws HoodieException {
-   * @throws HoodieException
-   */
-  public void stop() throws HoodieException {
-    instantToHeartbeatMap.values().forEach(heartbeat -> stop(heartbeat.getInstantTime()));
```
We need to at least stop timers for the existing heartbeats

stream2000 commented:
We should also note that if we don't restart the job before the heartbeat expires, the cleaner will still roll back the inflight instant that should be recommitted. And we can't set the heartbeat time too long because clean and archive will be blocked by inflight instants.

SteNicholas commented:
> We should also note that if we don't restart the job before the heartbeat expires, the cleaner will still roll back the inflight instant that should be recommitted. And we can't set the heartbeat time too long because clean and archive will be blocked by inflight instants.

@stream2000, this requires that the recommit operation should execute within `maxAllowableHeartbeatIntervalInMs`, otherwise the recommit instant will rollback for the clean action.

@stream2000, I have addressed above comments. PTAL.

stream2000 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -186,35 +180,54 @@ public void start(String instantTime) {
+   *
+   * @param heartbeat The heartbeat to stop.
+   */
+  public void stopHeartbeat(Heartbeat heartbeat) {
```
We don't need to make it public right?

stream2000 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -186,35 +180,54 @@ public void start(String instantTime) {
+   * @return Whether the heartbeat is started.
+   * @throws IOException
+   */
+  public boolean isHeartbeatStarted(Heartbeat heartbeat) {
```
ditto

stream2000 reviewed (approved):
LGTM

stream2000 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -185,36 +179,55 @@ public void start(String instantTime) {
 
   /**
    * Stops all heartbeats started via this instance of the client.
+   *
```
We can add a comment like 'Note that  we won't delete the heartbeat file by stop'  or just rename this method to `stopTimers`

SteNicholas commented:
@n3nash, could you help to review this pull request?

leesf commented:
LGTM, let's waiting for https://github.com/apache/hudi/pull/9163 to be merged to fix CI.

danny0405 commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/testutils/HoodieClientTestBase.java:
```diff
@@ -770,6 +770,7 @@ protected void updateBatchWithoutCommit(String newCommitTime, List<HoodieRecord>
         .withAutoCommit(false) // disable auto commit
         .withRollbackUsingMarkers(true)
+        .withHeartbeatTolerableMisses(0)
         .build();
```
What the purpose of this change?

SteNicholas replied:
@danny0405, accelerate the expiration of the heartbeat for the deltacommit and not wait for the heartbeat expiration in [line 308](https://github.com/apache/hudi/blob/master/hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/functional/TestSavepointRestoreMergeOnRead.java#L308) for `TestSavepointRestoreMergeOnRead#testCleaningCompletedRollback`.

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -185,36 +179,55 @@ public void start(String instantTime) {
+   * @throws IOException
+   */
+  private boolean isHeartbeatStarted(Heartbeat heartbeat) {
+    return heartbeat != null && heartbeat.isHeartbeatStarted() && !heartbeat.isHeartbeatStopped();
```
use `heartbeat != null && heartbeat.isHeartbeatStarted()` or `heartbeat != null && heartbeat.isHeartbeatStopped()` is enough?

SteNicholas replied:
@leesf, `heartbeat != null && heartbeat.isHeartbeatStarted()` or `heartbeat != null && heartbeat.isHeartbeatStopped()` is not enough. Because `isHeartbeatStarted` is true and `isHeartbeatStopped` is true when heartbeat is stopped.

danny0405 commented:
cc @bvaradar , can you take a look at this change? I'm not very clear with the impact on Spark engine/workflow.

SteNicholas commented:
@bvaradar, could you help to review this change?

bvaradar reviewed (commented):
LGTM.
